wiki/What-If.md

```diff
@@ -23,53 +32,619 @@ often are not. A non-exhaustive list:
+
+There is also quite a lot of 'research software' where the surface
+requirements change all the time, but the deep theory is essentially fixed.
+The ITD process really lets you do "what if" experiments in those settings
+quite easily.
+
+But of course, pretty much all of the software where Agile would be
+appropriate are domains where ITD would not be.
+
+**N** Oh yes, that makes a lot of sense.
+
+## Why now?
+
+Hmm, you mentioned a number of projects that were related (Draco, DMS, GLisp,
+SpecWare). But I didn't really
+hear about any of them. Did they fail? And what's different now?
+
+**S** Good point.
+
+It's not so much that they failed, but that the tools part of the projects
+did not survive. Many of the projects had a substantial intellectural
+impact on many subsequent projects.
+
+The main reason for the tools to fall to the wayside is that the maintenance
+burden was too high compared to the amount of take-up by outsiders in a
+reasonable time frame. And a lot of that had three underlying reasons:
+1) poor technology, 2) overly ambitious aims, and 3) lack of theoretical
+foundations. Any one of them is not a fatal flaw, but the combination of all
+of them was too much.
+
+**N** So you think that's changed?
+
+**S** Yes, absolutely. There have been tremendous advances in
+meta-programming, statically typed programming, tooling for DSLs and
+generally a much better understanding on how to create languages.
+Our project tries to mitigate the 'overly ambitious' by trying to do
+the minimal infrastructure necessary to get things to work, and then
+refactor when necessary. We're also very example-driven when it comes to
+assessing the need for infrastructure. In other words, our own internal
+development is really not following our own process -- basically because
+we're continually inventing, the very opposite of 'well understood'.
+Lastly, the theory underlying languages (programming and specification alike,
+but also natural languages as well) has also improved dramatically.
+
+Basically enough has changed between then and now that it feels worthwhile
+learning the lessons of the older projects (i.e. both their good ideas and
+their mistakes) to try again. Of course we're also in a good position to
+learn from many projects.
+
+**N** Well, I think that covers it.
+
+**S** You forget one of your questions about "semantic sources".
+
+**N** Oh yeah!
+
+## Semantic sources
+
+**S** You might have noticed that I called it 'Idealized Process for
+Theory-based Development' and talked about theories a lot, but never quite
+explained that.
+
+**N** Now that you mention it...
+
+**S** I was equally vague with 'knowledge', wasn't I? The point is that
+whatever information we have, it needs to be organized. So we use two
+different mechanisms: theories and ontologies.
+
+Theories in the mathematical sense (defining types, function and
+relation symbols, obeying some axioms) but also in the more general
```

Comment:
I don't think this needs to be added, but I thought I would note it: "mathematical sense" means more than "just numbers" (which you highlight in your examples). For example, a theory could be used to define the denotational semantics of a data type, which can mean source code.